state/protocol/state.go

```diff
@@ -37,25 +37,44 @@ type State interface {
 	AtBlockID(blockID flow.Identifier) Snapshot
 }
 
-type MutableState interface {
+// FollowerState is a mutable protocol state used by nodes following main consensus (ie. non-consensus nodes).
+// All blocks must have a certifying QC when being added to the state to guarantee they are valid,
+// so there is a one-block lag between block production and incorporation into the FollowerState.
+// However, since all blocks are certified upon insertion, they are immediately processable by other components.
+type FollowerState interface {
 	State
-	// Extend introduces the block with the given ID into the persistent
+	// ExtendCertified introduces the block with the given ID into the persistent
 	// protocol state without modifying the current finalized state. It allows
 	// us to execute fork-aware queries against ambiguous protocol state, while
 	// still checking that the given block is a valid extension of the protocol state.
-	// Depending on implementation it might be a lighter version that checks only block header.
-	// The candidate block must have passed HotStuff validation before being passed to Extend.
+	// Caller must pass a QC for candidate block to prove that candidate block
+	// has been certified, and it's safe to add it to the block state.
+	// QC cannot be nil and must certify candidate block (candidate.View == qc.View && candidate.BlockID == qc.BlockID)
 	// Expected errors during normal operations:
```

Comment:
```suggestion
	// The `candidate` block and its QC _must be valid_ (otherwise, the state will be corrupted).
	// Expected errors during normal operations:
```